datumaro/components/media.py

```diff
@@ -178,3 +179,276 @@ def save(self, path):
+        return VideoFrame(self._video, index=index)
+
+    def __next__(self):
+        return next(self._iterator)
+
+    def __getitem__(self, idx: int) -> VideoFrame:
+        if not self._video._includes_frame(idx):
+            raise IndexError(f"Video doesn't contain frame #{idx}.")
+
+        return self._navigate_to(idx)
+
+    def get_frame_data(self, idx: int) -> np.ndarray:
+        self._navigate_to(idx)
+        return self._current_frame_data
+
+    def _navigate_to(self, idx: int) -> VideoFrame:
+        """
+        Iterates over frames to the required position.
+        """
+
+        if idx < 0:
+            raise IndexError()
+
+        if idx < self._pos:
+            self._reset()
+
+        if self._pos < idx:
+            try:
+                while self._pos < idx:
+                    v = self.__next__()
+            except StopIteration as e:
+                raise IndexError() from e
+        else:
+            v = self._make_frame(index=self._pos)
+
+        return v
+
+class Video(MediaElement, Iterable[VideoFrame]):
+    """
+    Provides random access to the video frames.
+    """
+
+    def __init__(self, path: str, *,
+            step: int = 1, start_frame: int = 0,
+            end_frame: Optional[int] = None) -> None:
+        super().__init__(path)
+
+        if end_frame:
+            assert start_frame < end_frame
+        assert 0 < step
+        self._step = step
+        self._start_frame = start_frame
+        self._end_frame = end_frame or None
+
+        self._reader = None
+        self._iterator: Optional[_VideoFrameIterator] = None
+        self._frame_size: Optional[Tuple[int, int]] = None
+
+        self._frame_count = None
+        self._length = None
+
+    def close(self):
+        self._iterator = None
+
+        if self._reader is not None:
+            self._reader.release()
+            self._reader = None
+
+    def __getitem__(self, idx: int) -> VideoFrame:
+        if not self._includes_frame(idx):
+            raise IndexError(f"Video doesn't contain frame #{idx}.")
+
+        return self._get_iterator()[idx]
+
+    def get_frame_data(self, idx: int) -> VideoFrame:
+        if not self._includes_frame(idx):
+            raise IndexError(f"Video doesn't contain frame #{idx}.")
+
+        return self._get_iterator().get_frame_data(idx)
+
+    def __iter__(self) -> Iterator[VideoFrame]:
+        """
+        Iterates over frames lazily, if possible.
+        """
+
+        if self._frame_count is not None:
+            # Decoding is not necessary to get frame pointers
+            # However, it can be inacurrate
+            end_frame = self._get_end_frame()
+            for index in range(self._start_frame, end_frame, self._step):
+                yield VideoFrame(video=self, index=index)
+        else:
+            # Need to decode to iterate over frames
+            yield from self._get_iterator()
+
+    @property
+    def est_frame_count(self) -> Optional[int]:
```

Comment:
IMO, this should either be renamed to `est_total_frame_count` (with appropriate updates to the documentation), or just not provided at all. Otherwise, it's too similar to `length`. Personally, I would argue for not providing it, because the `Video` class essentially represents a _segment_ of a video, and the total length of the video shouldn't be relevant.

Reply:
I removed the function until we get more understanding on it's use.